pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -395,145 +405,161 @@ func (service *SecurityPolicyService) createOrUpdateSecurityPolicy(obj *v1alpha1
-		}
+	securityPolicyStore, ruleStore, groupStore, projectGroupStore, shareStore := service.getStores()
+	indexScope := common.TagValueScopeSecurityPolicyUID
+	if createdFor == common.ResourceTypeNetworkPolicy {
+		indexScope = common.TagScopeNetworkPolicyUID
+	}
 
-		// 2.Wrap SecurityPolicy, groups, rules under VPC level together with project groups and shares into one hierarchy resource tree.
-		orgRoot, err := service.WrapHierarchyVpcSecurityPolicy(finalSecurityPolicy, finalGroups, projectInfra, vpcInfo)
-		if err != nil {
-			log.Error(err, "failed to wrap SecurityPolicy in VPC")
-			return err
-		}
+	// WrapHierarchyVpcSecurityPolicy will modify the input security policy, so we need to make a copy for the following store update.
+	finalSecurityPolicyCopy := *finalSecurityPolicy
 
-		// 3.Create/update SecurityPolicy together with groups, rules under VPC level and project groups, shares.
-		err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam)
-		err = nsxutil.NSXApiError(err)
-		if err != nil {
-			log.Error(err, "failed to create or update SecurityPolicy in VPC")
-			return err
-		}
+	nsxProjectGroups := make([]model.Group, 0)
+	nsxProjectShares := make([]model.Share, 0)
+	for i := len(*projectShares) - 1; i >= 0; i-- {
+		nsxProjectGroups = append(nsxProjectGroups, *((*projectShares)[i].shareGroup))
+		nsxProjectShares = append(nsxProjectShares, *((*projectShares)[i].share))
+	}
 
-		if len(finalProjectGroups) != 0 {
-			err = projectGroupStore.Apply(&finalProjectGroups)
-			if err != nil {
-				log.Error(err, "failed to apply store", "nsxProjectGroups", finalProjectGroups)
-				return err
-			}
-		}
+	// Create/Update nsx project shares and nsx project level groups
+	existingNsxProjectGroups := projectGroupStore.GetByIndex(indexScope, string(obj.UID))
+	finalProjectGroups := service.updateGroups(existingNsxProjectGroups, nsxProjectGroups)
 
-		if len(finalProjectShares) != 0 {
-			err = shareStore.Apply(&finalProjectShares)
-			if err != nil {
-				log.Error(err, "failed to apply store", "nsxProjectShares", finalProjectShares)
-				return err
-			}
-		}
-	} else {
-		infraSecurityPolicy, err := service.WrapHierarchySecurityPolicy(finalSecurityPolicy, finalGroups)
-		if err != nil {
-			log.Error(err, "failed to wrap SecurityPolicy")
-			return err
-		}
-		err = service.NSXClient.InfraClient.Patch(*infraSecurityPolicy, &EnforceRevisionCheckParam)
-		err = nsxutil.NSXApiError(err)
+	existingNsxProjectShares := shareStore.GetByIndex(indexScope, string(obj.UID))
+	finalProjectShares := service.updateShares(existingNsxProjectShares, nsxProjectShares)
+
+	if !isChanged && len(finalSecurityPolicy.Rules) == 0 && len(finalGroups) == 0 && len(finalProjectGroups) == 0 && len(finalProjectShares) == 0 {
+		log.Info("securityPolicy, rules, groups and shares are not changed, skip updating them", "nsxSecurityPolicy.Id", finalSecurityPolicy.Id)
+		return nil
+	}
+
+	// 1.Wrap project groups and shares into project child infra.
```

Comment:
L533-L556 and L753-L776, L568-L595 and L778-L802 are also dup, which can be simplified... Can you add a TODO comment for it? We can polish them in the future..

Reply:
sure, we need to polish the code further to remove duplicated code, I gonna to do it in the next patch.